### PR TITLE
[e2e] Scheduler priority test

### PR DESCRIPTION
**Pull request by wanghaoran1988**
Fix https://github.com/kubernetes/kubernetes/issues/42439
In order to reduce the interfere of the balanced least-requested priority function, I create a balance pod to make the node have same mem/cpu use rate, not sure it's a correct way, but it did good, I have not cover all the scenarios yet, want to get feedback whether I am in the right way.
@k82cn @bsalamat Would you please help review ?

### CODE REVIEW

**Comment by bsalamat** [comment]
@k8s-bot ok to test

**Comment by k82cn** [comment]
@wanghaoran1988 , thanks for your PR :). I'll take some to review it this or next week.

**Comment by k82cn** [comment]
@kubernetes/sig-scheduling-pr-reviews 

Propose to be **1.7** candidates.

**Comment by timothysc** [comment]
@wanghaoran1988 per-policy please /cc @kubernetes/sig-scheduling-pr-reviews in case folks decide to go on vacation etc.

**Comment by wanghaoran1988** [comment]
@timothysc Seems cannot /cc the private group :)

File: test/e2e/scheduler_priorities.go [hunk 1]
```diff
@@ -0,0 +1,715 @@
...
+			result[pod.Spec.NodeName] = true
+		}
+		By("Verify the pods were scheduled to each node")
+		if len(nodeList.Items) != len(result) {
+
+			framework.Failf("Pods are not spread to each node")
+		}
+
+	})
+	It("Pods Belong to same StatefulSet should spread to different node", func() {
```
Comment (bsalamat) [inline test/e2e/scheduler_priorities.go]: nit: Please add a blank line before the test and remove the one after

File: test/e2e/scheduler_priorities.go [hunk 2]
```diff
@@ -0,0 +1,715 @@
...
+		framework.ExpectNodeHasAvoidPods(cs, nodeName, avoidPod)
+		defer framework.RemoveAvoidPodsOffNode(cs, nodeName)
+
+		By(fmt.Sprintf("Scale the RC: %s to len(nodeList.Iten)-1 : %v.", rc.Name, len(nodeList.Items)-1))
+
+		framework.ScaleRC(f.ClientSet, f.InternalClientset, ns, rc.Name, uint(len(nodeList.Items)-1), true)
+		testPods, err := cs.Core().Pods(ns).List(metav1.ListOptions{
+			LabelSelector: "name=scheduler-priority-avoid-pod",
+		})
+		framework.PrettyPrintJSON(testPods)
```
Comment (bsalamat) [inline test/e2e/scheduler_priorities.go]: Is this line needed?

File: test/e2e/scheduler_priorities.go [hunk 3]
```diff
@@ -0,0 +1,715 @@
...
+		//we need applay more taint on a node, because one match toleration only count 1
+		By("Trying to apply 10 taint on the nodes except first one.")
+		nodeName := nodeList.Items[0].Name
+
+		for index, node := range nodeList.Items {
+			if index == 0 {
+				continue
+			}
+			for i := 0; i < 10; i++ {
+				testTaint := v1.Taint{
```
Comment (bsalamat) [inline test/e2e/scheduler_priorities.go]: Optional: I would create a function to for this block, as it is repeated below and may be useful in the future, but up to you.

**Review by bsalamat — CHANGES_REQUESTED** [review]
First of all sorry for the delay in reviewing your PR and thanks a lot for taking the time to write the tests. This is very valuable and I really appreciate it. As our leads also said in a separate discussion, we should all try harder to improve our code coverage.
I have a few comments which are mostly nits.

**Comment by wanghaoran1988** [comment]
@k8s-bot unit test this

**Comment by wanghaoran1988** [comment]
@bsalamat Thanks for your comments, they are addressed now, PTAL.

**Comment by bsalamat** [comment]
Review status: 0 of 4 files reviewed at latest revision, 23 unresolved discussions.

---

*[test/e2e/framework/util.go, line 2663 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KfgxOxG6_Y4Ei6DP3Vi:-KfgxOxG6_Y4Ei6DP3Vj:b-fk6n8q) ([raw file](https://github.com/kubernetes/kubernetes/blob/5bbc86a640ddcd9cef93950aca2b318f5087398d/test/e2e/framework/util.go#L2663)):*
> ```Go
> 			}
> 		}
> 		time.Sleep(100 * time.Millisecond)
> ```

Why the hard sleep is needed here? Hard sleeps usually introduce flakes. Can we remove/replace it?

---

*[test/e2e/scheduling/priorities.go, line 161 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-Kfgy9ZbWzG2JKafxzOm:-Kfgy9ZbWzG2JKafxzOn:b-56xtix) ([raw file](https://github.com/kubernetes/kubernetes/blob/5bbc86a640ddcd9cef93950aca2b318f5087398d/test/e2e/scheduling/priorities.go#L161)):*
> Pods NO. are not corrected
Unexpected number of pods, ...

---

*[test/e2e/scheduling/priorities.go, line 517 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-Kfh1TNlcsbABQjTR0OQ:-Kfh1TNlcsbABQjTR0OR:b-zi3ant) ([raw file](https://github.com/kubernetes/kubernetes/blob/5bbc86a640ddcd9cef93950aca2b318f5087398d/test/e2e/scheduling/priorities.go#L517)):*
> ```Go
> })
> 
> //This function create pod on each node to request enough resource to make the nodes have same mem/cpu usage ratio
> ```

// createBalancedPodForNodes creates a pod per node that asks for enough resources to make all nodes have the same mem/cpu usage ratio.

---

*[test/e2e/scheduling/priorities.go, line 692 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-Kfh0sfZY3-0tG1Ij6rY:-Kfh0sf_Wbc634URwK5j:b-dahnp4) ([raw file](https://github.com/kubernetes/kubernetes/blob/5bbc86a640ddcd9cef93950aca2b318f5087398d/test/e2e/scheduling/priorities.go#L692)):*
> v1.Taint

This is a test and we probably don't care much about performance and avoiding copies, but as a good practice, we should probably return a pointer: *v1.Taint

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890)*

**Comment by bsalamat** [comment]
Thanks again, Haoran! We are almost there.

---

Review status: 0 of 4 files reviewed at latest revision, 23 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-:-Kfh1po4PtBvBdBW2F6m:b-j3slzs)*

**Comment by wanghaoran1988** [comment]
@bsalamat The comments are addressed, thanks for your comments and careful review, and the sentences you correct for me, my English is not good, but I am on my way learning English. :)

**Comment by wanghaoran1988** [comment]
Review status: 0 of 4 files reviewed at latest revision, 23 unresolved discussions.

---

*[test/e2e/framework/util.go, line 2607 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZMJ-r1-2607:-KfjVeUN_TDYIcsv0hAh:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/framework/util.go#L2607)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

Core() is deprecated. Please use CoreV1() for new code (here and the rest of the PR).
</blockquote></details>

Done.

---

*[test/e2e/framework/util.go, line 2629 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZMJ-r1-2629:-KfjqGBrua1XmsSAphxJ:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/framework/util.go#L2629)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

Please remove the blank line.
</blockquote></details>

Done.

---

*[test/e2e/framework/util.go, line 2632 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZMJ-r1-2632:-KfjqH8LN1H7mOyTaxg8:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/framework/util.go#L2632)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

This function removes all the AvoidPods annotations regardless of their temporary lifespan. So, the following might better reflect what it does:

// RemoveAnnotationOffNode removes AvoidPods annotations from the node. It does not fail if no such annotation exists.
</blockquote></details>

Done.

---

*[test/e2e/framework/util.go, line 2655 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZMJ-r1-2655:-KfjqJhjazZrNnOJKru_:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/framework/util.go#L2655)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

Please remove the blank line.
</blockquote></details>

Done.

---

*[test/e2e/framework/util.go, line 2663 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KfgxOxG6_Y4Ei6DP3Vi:-KfjqILHoHnu93-XtMG6:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/5bbc86a640ddcd9cef93950aca2b318f5087398d/test/e2e/framework/util.go#L2663)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

Why the hard sleep is needed here? Hard sleeps usually introduce flakes. Can we remove/replace it?
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 1 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZML-r1-1:-KfjW6FsUU8FEcgeAM-Y:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/scheduling/priorities.go#L1)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

After the recent changes, you should move the file to test/e2e/scheduling/priorities.go
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 71 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZML-r1-71:-KfjW337m28q2Vfomw2L:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/scheduling/priorities.go#L71)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

nit: Please remove the blank line. There are a few other cases throughout the file where blank lines are added right after "{" or before "}". Please remove all of them.
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 90 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZML-r1-90:-KfjVyoElZOtG6HYFnL6:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/scheduling/priorities.go#L90)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

nit: `// Make sure all the schedulable nodes are balanced (have the same cpu/mem usage ratio)`
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 91 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZML-r1-91:-KfjW-gFWSuf7f7rc_6z:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/scheduling/priorities.go#L91)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

nit: `By("Create pods on each node except the last one to raise cpu and memory usage to the same high level")`
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 118 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZML-r1-118:-KfjW5pGUUozZFRxAZPl:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/scheduling/priorities.go#L118)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

Sorry for being pedantic. Feel free to ignore my nit comments. I just try to make the sentences a bit more clear.

nit: By("Create an RC with the same number of replicas as the schedualble nodes. One pod should be scheduled on each node.")
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 166 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZML-r1-166:-KfjW6qo3ZDVYq4r_sY4:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/scheduling/priorities.go#L166)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

Please add a check: len(pods) == len(nodeList.Items)
This is not strictly needed, but will help find the problem before the check at line 179 fails.
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 185 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZML-r1-185:-KfjW7FsXYimsNFtB1nO:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/scheduling/priorities.go#L185)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

Please add a blank line before this test! :-)

nit: Pods of the same service should spread to different nodes.
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 190 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZML-r1-190:-KfjW888K99AF63JyFbR:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/scheduling/priorities.go#L190)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

Please remove // from the end
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 191 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZML-r1-191:-KfjW8ZaQz1u90P4B6sU:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/scheduling/priorities.go#L191)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

Similar nit as the test above.
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 225 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZML-r1-225:-KfjW8xO01ddwwEg0mgU:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/scheduling/priorities.go#L225)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

nit: Please add a blank line before the test and remove the one after
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 473 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZML-r1-473:-KfjW9KijmFrrw1oX3Yy:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/scheduling/priorities.go#L473)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

Is this line needed?
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 492 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZML-r1-492:-KfjWBKTyqontsiLaHir:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/scheduling/priorities.go#L492)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

Optional: I would create a function to for this block, as it is repeated below and may be useful in the future, but up to you.
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 542 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZML-r1-542:-KfjW4jWCt6ZJoqFLWaF:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/scheduling/priorities.go#L542)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

Please add a comment to explain that this function creates pods requesting enough resources so that the same 'ratio' of resources are consumed on the 'nodes'.
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 608 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZML-r1-608:-KfjW3b5qVFzyER3QZhu:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/scheduling/priorities.go#L608)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

Please remove the blank line
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 619 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KesvdDx9pfzGb9B3ZML-r1-619:-KfjW4JDisptH-N-dLST:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/780625caa1c0699c138bea058d6e5f043b1afb5d/test/e2e/scheduling/priorities.go#L619)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

Please remove the blank line
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 161 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-Kfgy9ZbWzG2JKafxzOm:-KfjWBwE66Lt43gg7SwP:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/5bbc86a640ddcd9cef93950aca2b318f5087398d/test/e2e/scheduling/priorities.go#L161)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

> Pods NO. are not corrected
Unexpected number of pods, ...

</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 517 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-Kfh1TNlcsbABQjTR0OQ:-KfjqTnpuWrEci9Zu48i:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/5bbc86a640ddcd9cef93950aca2b318f5087398d/test/e2e/scheduling/priorities.go#L517)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

// createBalancedPodForNodes creates a pod per node that asks for enough resources to make all nodes have the same mem/cpu usage ratio.
</blockquote></details>

Done.

---

*[test/e2e/scheduling/priorities.go, line 692 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-Kfh0sfZY3-0tG1Ij6rY:-KfjqW06XZindhk3b-Oq:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/5bbc86a640ddcd9cef93950aca2b318f5087398d/test/e2e/scheduling/priorities.go#L692)):*
<details><summary><i>Previously, bsalamat (Bobby (Babak) Salamat) wrote…</i></summary><blockquote>

> v1.Taint

This is a test and we probably don't care much about performance and avoiding copies, but as a good practice, we should probably return a pointer: *v1.Taint
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890)*

**Comment by wanghaoran1988** [comment]
@k8s-bot unit test this

**Comment by bsalamat** [comment]
Review status: 0 of 4 files reviewed at latest revision, 24 unresolved discussions.

---

*[test/e2e/scheduling/priorities.go, line 515 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890#-KfnPqeMSeH1vvxLEX8l:-KfnPqeMSeH1vvxLEX8m:b-8djzj3) ([raw file](https://github.com/kubernetes/kubernetes/blob/259abaa0ae9037536e11bc5140f5ac9d29a999a7/test/e2e/scheduling/priorities.go#L515)):*
> Create a pod tolerate all the taits of the first node
Create a pod that tolerates all the taints of the first node.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42890)*

**Comment by wanghaoran1988** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by bsalamat** [comment]
Thanks again, Haoran!

File: test/e2e/scheduling/priorities.go [hunk 4]
```diff
@@ -0,0 +1,713 @@
...
+		//Make sure all the schedulable nodes are balanced (have the same cpu/mem usage ratio)
+		By("Create pods on each node except the last one to raise cpu and memory usage to the same high level")
+		var expectedNodeName string
+		expectedNodeName = nodeList.Items[len(nodeList.Items)-1].Name
+		nodes := nodeList.Items[:len(nodeList.Items)-1]
+		// make the nodes except last have cpu,mem usage to 90%
+		createBalancedPodForNodes(f, cs, ns, nodes, podRequestedResource, 0.9)
+		By("Create a pod,pod should schedule to the least requested nodes")
+		createPausePod(f, pausePodConfig{
+			Name:      "priority-least-requested-mem",
```
Comment (k82cn) [inline test/e2e/scheduling/priorities.go]: s/priority-least-requested-mem/priority-least-requested/

as both cpu and mem are included.

File: test/e2e/scheduling/priorities.go [hunk 5]
```diff
@@ -0,0 +1,713 @@
...
+		By("Wait for all the pods are running")
+		err := f.WaitForPodRunning("priority-least-requested-mem")
+		framework.ExpectNoError(err)
+		By("Verify the pod is scheduled to the expected node")
+		testPod, err := cs.Core().Pods(ns).Get("priority-least-requested-mem", metav1.GetOptions{})
+		framework.ExpectNoError(err)
+		Expect(testPod.Spec.NodeName).Should(Equal(expectedNodeName))
+	})
+
+	It("Pods created by ReplicationController should spread to different node", func() {
```
Comment (k82cn) [inline test/e2e/scheduling/priorities.go]: I think one case for `ReplicationController`, `ReplicaSet`, `StatefulSet` and `Service` is enough, as the scheduling logic is same; the only different is how to get selector (we can cover it in unit test.)

Less cases will not impact the coverage; but reduce e2e time.

File: test/e2e/scheduling/priorities.go [hunk 6]
```diff
@@ -0,0 +1,713 @@
...
+		createBalancedPodForNodes(f, cs, ns, nodeList.Items, podRequestedResource, 0.6)
+		By("Trying to relaunch the pod, now with labels.")
+		labelPodName := "pod-with-node-affinity"
+		pod := createPausePod(f, pausePodConfig{
+			Name: labelPodName,
+			Affinity: &v1.Affinity{
+				NodeAffinity: &v1.NodeAffinity{
+					PreferredDuringSchedulingIgnoredDuringExecution: []v1.PreferredSchedulingTerm{
+						{
+							Preference: v1.NodeSelectorTerm{
```
Comment (k82cn) [inline test/e2e/scheduling/priorities.go]: Why two selector here (hostname + k=v)?

File: test/e2e/scheduling/priorities.go [hunk 7]
```diff
@@ -0,0 +1,713 @@
...
+			Name:      "with-label-security-s1",
+			Labels:    map[string]string{"service": "S1"},
+			Resources: podRequestedResource,
+		})
+		nodeName := pod.Spec.NodeName
+
+		By("Trying to apply label on the found node.")
+		k := fmt.Sprintf("kubernetes.io/e2e-%s", "node-topologyKey")
+		v := "topologyvalue"
+		framework.AddOrUpdateLabelOnNode(cs, nodeName, k, v)
```
Comment (k82cn) [inline test/e2e/scheduling/priorities.go]: Why add label to node? It's testing pod affinity.

File: test/e2e/scheduling/priorities.go [hunk 8]
```diff
@@ -0,0 +1,713 @@
...
+			Name:   "pod-with-label-security-s1",
+			Labels: map[string]string{"security": "S1"},
+		})
+		nodeName := pod.Spec.NodeName
+
+		By("Trying to apply a label on the found node.")
+		k := fmt.Sprintf("kubernetes.io/e2e-%s", "node-topologyKey")
+		v := "topologyvalue"
+		framework.AddOrUpdateLabelOnNode(cs, nodeName, k, v)
+		framework.ExpectNodeHasLabel(cs, nodeName, k, v)
```
Comment (k82cn) [inline test/e2e/scheduling/priorities.go]: ditto

File: test/e2e/scheduling/priorities.go [hunk 9]
```diff
@@ -0,0 +1,713 @@
...
+			}
+			return node.Annotations[v1.PreferAvoidPodsAnnotationKey] == string(val)
+		}
+		success, err := common.ObserveNodeUpdateAfterAction(f, nodeName, predicate, action)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(success).To(Equal(true))
+
+		defer framework.RemoveAvoidPodsOffNode(cs, nodeName)
+
+		By(fmt.Sprintf("Scale the RC: %s to len(nodeList.Iten)-1 : %v.", rc.Name, len(nodeList.Items)-1))
```
Comment (k82cn) [inline test/e2e/scheduling/priorities.go]: s/Iten/Items

File: test/e2e/scheduling/priorities.go [hunk 10]
```diff
@@ -0,0 +1,713 @@
...
+
+		By("Pod should prefer scheduled to the node that pod can tolerate.")
+		tolePod, err = cs.Core().Pods(ns).Get(tolerationPodName, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(tolePod.Spec.NodeName).To(Equal(nodeName))
+	})
+})
+
+// createBalancedPodForNodes creates a pod per node that asks for enough resources to make all nodes have the same mem/cpu usage ratio.
+func createBalancedPodForNodes(f *framework.Framework, cs clientset.Interface, ns string, nodes []v1.Node, requestedResource *v1.ResourceRequirements, ratio float64) {
```
Comment (k82cn) [inline test/e2e/scheduling/priorities.go]: Why we need this func? Suppose all nodes has similar allocatable.

**Review by k82cn — CHANGES_REQUESTED** [review]

File: test/e2e/framework/util.go [hunk 11]
```diff
@@ -2615,6 +2616,55 @@ func NodeHasTaint(c clientset.Interface, nodeName string, taint *v1.Taint) (bool
 	return true, nil
 }
 
+//AddOrUpdateAvoidPodOnNode add avoidPods annotations to node, will override if it exists
+func AddOrUpdateAvoidPodOnNode(c clientset.Interface, nodeName string, avoidPods v1.AvoidPods) {
+	err := wait.PollImmediate(Poll, SingleCallTimeout, func() (bool, error) {
+		node, err := c.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
```
Comment (wanghaoran1988) [inline test/e2e/framework/util.go]: @k82cn Thanks, will update all to use CoreV1() per comment here: https://github.com/kubernetes/kubernetes/pull/42890#discussion_r105520490

File: test/e2e/scheduling/priorities.go (hunk 5 above)
Comment (wanghaoran1988) [inline test/e2e/scheduling/priorities.go]: @k82cn I just want to make sure all of them are covered

File: test/e2e/scheduling/priorities.go (hunk 7 above)
Comment (wanghaoran1988) [inline test/e2e/scheduling/priorities.go]: @k82cn For the topologyKey in the podAffinity iterm

Comment (k82cn) [inline test/e2e/scheduling/priorities.go]: Yes, prefer to add two nodes in this topology; current case is "similar" to node affinity.

File: test/e2e/scheduling/priorities.go (hunk 5 above)
Comment (k82cn) [inline test/e2e/scheduling/priorities.go]: prefer to just keep one case which is enough to cover the "selector" priority's logic; for how to get selector, I'd like to dependent on unit test :).

**Comment by k82cn** [comment]
@wanghaoran1988 , please help to address comments, and then we can merge this PR :).

File: test/e2e/scheduling/priorities.go (hunk 5 above)
Comment (wanghaoran1988) [inline test/e2e/scheduling/priorities.go]: @k82cn Yeah, sounds reasonable, I can remove it.

File: test/e2e/scheduling/priorities.go (hunk 6 above)
Comment (wanghaoran1988) [inline test/e2e/scheduling/priorities.go]: @k82cn No, just add two item to match the affinity, but we can using one selector if you don't like.

File: test/e2e/scheduling/priorities.go (hunk 7 above)
Comment (wanghaoran1988) [inline test/e2e/scheduling/priorities.go]: @k82cn nodeAffinity don't have this key, this is just use to filter out the pods that on the node with that label

File: test/e2e/scheduling/priorities.go (hunk 10 above)
Comment (wanghaoran1988) [inline test/e2e/scheduling/priorities.go]: Similar does not mean same, as before we run the test, there really have a lot system pods make the resource not balanced.

File: test/e2e/scheduling/priorities.go (hunk 6 above)
Comment (k82cn) [inline test/e2e/scheduling/priorities.go]: I'm OK with that.

File: test/e2e/scheduling/priorities.go (hunk 10 above)
Comment (k82cn) [inline test/e2e/scheduling/priorities.go]: Can we remove pods after each cases?

Comment (wanghaoran1988) [inline test/e2e/scheduling/priorities.go]: @k82cn The namespaces will be deleted so the pods will be deleted also.

**Comment by wanghaoran1988** [comment]
@k82cn Comments are addressed, would you mind add lgtm lable?

File: test/e2e/scheduling/priorities.go (hunk 10 above)
Comment (k82cn) [inline test/e2e/scheduling/priorities.go]: I means deleting Pods after each test cases (If the pods was deleted, why we need this?). 

The ratio seems different from cases, e.g. 0.5, 0.6, 0.9, how did you define it? and how other contributor define it for other e2e cases?

File: test/e2e/scheduling/priorities.go (hunk 7 above)
Comment (k82cn) [inline test/e2e/scheduling/priorities.go]: I'm OK with that; not a block comments :).

File: test/e2e/scheduling/priorities.go (hunk 10 above)
Comment (wanghaoran1988) [inline test/e2e/scheduling/priorities.go]: @k82cn Every test case will have a new namespace before start and deleted after finished. the balanced pods are created for each cases, and will be deleted after each case. For the ratio, just set the value you want to is ok.

**Comment by k82cn** [comment]
/lgtm

Echo bsalamat's lgtm, although I have concern on `createBalancedPodForNodes`: `ratio` seems magic number to me: it need to know e2e environment (static pods), test cases and code of `createBalancedPodForNodes`. IMO, it's better to run those test on no-master/management nodes to avoid impact of static pods.

**Pull request merged**